### ISSUE TITLE
Native Open Policy Agent / Rego support in OpenTF

### ISSUE THREAD

**Issue by marcinwyszynski** [opentffoundation/opentf#370]
### Summary

I would like to propose adding native integration with [Rego](https://www.openpolicyagent.org/docs/latest/policy-reference/) to OpenTF, allowing Rego code to be directly called from OpenTF code.

### Problem Statement

OpenTF inherits from Terraform various [condition checks](https://developer.hashicorp.com/terraform/language/expressions/custom-conditions) but using them in the HCL-based language is cumbersome if non-trivial conditions are involved.

Also, more complex data structures transformations can be very cumbersome given the limited expressive power of the HCL-based language.

### User-facing description

I propose adding a simple function (eg. called `rego`) that accepts 3 inputs:

- a Rego query, eg `data.opentf.allow`;
- a Rego policy body;
- an arbitrary HCL input;

The result of this function would be a HCL representation of the result of the query. The return type would be dynamic, similar to how `jsondecode` now works.

### Technical Description

The change required is minimal because it just involves adding a single function to the list of currently supported ones.
A quick PoC is available [here](https://github.com/opentffoundation/opentf/pull/368).

### Rationale and alternatives

The impact of this feature allows native building proper policy-as-code guardrails into OpenTF without the need to involve external tools. It also allows the users to accomplish more complex data transformations in OpenTF that would be possible (and sensible) without a more specialized tool.

### Downsides

The size of the OpenTF binary will increase (about 10% on my local machine) because of an added non-trivial dependency.
Since the function does not exist in the legacy Terraform, users who adopt it will need to change their code if they wish to use the legacy option.

### Proof of Concept

The experimental PR is [here](https://github.com/opentffoundation/opentf/pull/368)

Upon building the binary, you can start it with the following config file:

```terraform
variable "keeper" {
  type    = string
  default = "foodsadfadsfad"
}

locals {
  policy = file("${path.module}/policy.rego")
}

resource "random_pet" "this" {
  keepers = {
    foo = var.keeper
  }

  lifecycle {
    precondition {
      condition     = rego("data.opentf.allow", local.policy, { id = var.keeper })
      error_message = join(", ", rego("data.opentf.errors", local.policy, { id = var.keeper }))
    }

    postcondition {
      condition     = rego("data.opentf.allow", local.policy, self)
      error_message = join(", ", rego("data.opentf.errors", local.policy, self))
    }
  }
}
```

The Rego file in question is:

```rego
package opentf

default allow = false

allow { count(errors) == 0 }

errors[sprintf("Resource id %q too long (%d)", [input.id, count(input.id)])] {
    print(sprintf("input: %s", [input]))
    count(input.id) > 5
}

errors["No way!"] { true }
```

Feel free to turn off and on postconditions and preconditions, and change the Rego file as needed.

**Comment by Hefeweizen** [comment]
Thanks for the proof-of-concept, it helps me visualize how to use this.  Two thoughts:

1. Rego was mentioned in the context of [cidrcontains](https://github.com/opentffoundation/opentf/issues/334); how would we write the cidrcontains test in rego?

I claim no real familiarity with Rego, so I'm using your sample:
```
data "aws_subnet" "containing" {
  id = "magic_subnet_id"
}

resource "aws_eip" "this" {

  lifecycle {
    postcondition {
      condition     = rego("data.opentf.cidr_contains", local.policy, { containing = data.aws_subnet.containing.cidr_block, contained = self.address })
      error_message = join(", ", rego("data.opentf.errors", local.policy, { id = var.keeper }))
    }
  }
}
```
The Rego policy file:
```
package opentf

default cidr_contains = false

cidr_contains { count(errors) == 0 }

errors[sprintf("New Ip not inside specified subnet")] {
    net.cidr_contains(input.containing, input.contained)
}

errors["No way!"] { true }
```

2. There's always a trade-off between expressiveness and conciseness.  This brings in the capability, but usage is verbose.  Can this be made simpler to use?

This simple policy has eight lines, and putting it inline in our local.policy would probably still be legible, but if it got longer, then the external file, as you have, feels almost necessary.  With multiple files, though, means that more effort is expended to read and write the code. Does it make any sense to have a Rego resource?  attached to other resources with a depends_on?

Similarly, would a data.rego_policy component create any readability?  I'm having to manually create a map of values (e.g. { id = var.keeper } in the provided poc; in mine, { containing = data.aws_subnet.containing.cidr_block, contained = self.address }).  Would a data block allow us to use these references directly?

In the end, I'm intrigued about adding this powerful feature to the tooling, but in its present form it seems like there will be lots of boilerplate.  That said, a simple function seems like a good place to start, further convenience could be added in the future, if the feature gains adoption.

**Comment by avnerenv0** [comment]
@marcinwyszynski why does this belong in the core?
Why not just have a "rego/opa provider" which can expose a data resource that does this?

**Comment by eranelbaz** [comment]
Why do we need this Rego support when we have `opentf test` feature?
Doesn't it makes us double implement Open Policy Agent? sounds for me a bit out of scope for the core project

**Comment by marcinwyszynski** [comment]
@avnerenv0 Yeah I think it's a valid alternative. Putting it in a provider would also provide much more flexibility in terms of versioning, adding new functionalities, supporting different approaches to querying (eg. bindings vs. expressions) etc.

**Comment by marcinwyszynski** [comment]
I need to think about it a bit more, but there's been a [long open PR](https://github.com/hashicorp/terraform/pull/31225) to legacy Terarform (by @apparentlymart), advocating for provider-defined functions, which could provide a good middle ground between core functions (relatively few, and all very stable) and external interfaces like data sources (easily defined and easily versioned, but more clunky to use).

**Comment by marcinwyszynski** [comment]
@eranelbaz 

> Why do we need this Rego support when we have opentf test feature?

These don't necessarily overlap. Rego can be used as a general purpose data transformation language. One use case is for making checks (like policy checks, but not exclusively) more readable and allowing for more complex logic. The other is to turn one data structure into another, or combining data multiple data structures in a complex way.

> Doesn't it makes us double implement Open Policy Agent?

Not necessarily, the concept is just about providing convenient access to the power of the underlying data query language, not duplicating the CLI capabilities.

> sounds for me a bit out of scope for the core project

This may well be true. I can see downsides. I think it's a very valid approach to keep the core minimal, and then make it simpler to externally supply such extensions, like mentioned above.

**Comment by eranelbaz** [comment]
Thank you for the detailed answer @marcinwyszynski  :)

**Comment by Hefeweizen** [comment]
@marcinwyszynski You closed the PR.  LMK if you'd like some help reworking this as a provider; I'd be glad to contribute.

**Comment by cube2222** [comment]
We've decided that this should be done as an external provider datasource, or a future provider function.
